### PR TITLE
Issue 9029

### PR DESCRIPTION
**Pull request by alimirjamali**
fixes: https://github.com/QubesOS/qubes-issues/issues/9029

### CODE REVIEW

File: qui/utils.py [hunk 1]
```diff
@@ -80,11 +80,25 @@ def run_asyncio_and_show_errors(loop, tasks, name, restart=True):
             exit_code = 1
     return exit_code
 
+def check_update(vm) -> bool:
```
Comment (marmarta) [inline qui/utils.py:83]: Everywhere this and check_support is used, they are wrapped in catching exceptions. I wonder if it wouldn't be neater to get this exception-catching done here, with some good defaults (no updates available, still supported?)

File: qui/updater/updater.py [hunk 2]
```diff
@@ -418,6 +419,12 @@ def parse_args(args, app):
                              'Interaction will be required in the event '
                              'of an update error.')
 
+    parser.add_argument('--show-advanced-settings', action='store_true',
```
Comment (marmarta) [inline qui/updater/updater.py:422]: See comment here: https://github.com/QubesOS/qubes-issues/issues/9029#issuecomment-2435510421

**Review by marmarta — CHANGES_REQUESTED** [review]

File: qui/updater/updater.py (hunk 2 above)
Comment (alimirjamali) [reply qui/updater/updater.py]: This is done. The updater will show the advanced settings by default and no CLI argument is needed.

File: qui/utils.py (hunk 1 above)
Comment (alimirjamali) [reply qui/utils.py:83]: > Everywhere this and check_support is used, they are wrapped in catching exceptions. I wonder if it wouldn't be neater to get this exception-catching done here, with some good defaults (no updates available, still supported?)

Honestly I do not remember why I did this. It was during the (relatively big & complex) global clipboard configurable size and `.metadata` project. And that one consumed a lot of brain power. I am going to move the catching exceptions inside them now. But I will make it a new commit in this PR for easier follow-up.

Comment (alimirjamali) [reply qui/utils.py:83]: This is done

**Review by marmarta — CHANGES_REQUESTED** [review]

File: qui/utils.py [hunk 3]
```diff
@@ -80,21 +80,41 @@ def run_asyncio_and_show_errors(loop, tasks, name, restart=True):
...
+    """Return true if the given template/standalone vm is updated or not
+    updateable or skipped. default returns true"""
+    try:
+        if not vm.features.get('updates-available', False):
+            return True
+        if not getattr(vm, 'updateable', False):
+            return True
+        if bool(vm.features.get('skip-update', False)):
+            return True
+    except exc.QubesException:
```
[the remark this answers is not in the document]
Comment (alimirjamali) [reply qui/utils.py:95]: Done

**Review by marmarta — APPROVED** [review]

File: qui/tray/updates.py [hunk 4]
```diff
@@ -126,44 +126,42 @@ def check_vms_needing_update(self):
...
+        self.dispatcher.add_handler('domain-feature-delete:skip-update',
+                                    self.feature_change)
         self.dispatcher.add_handler('domain-add', self.domain_added)
         self.dispatcher.add_handler('domain-delete', self.domain_removed)
         self.dispatcher.add_handler('domain-feature-set:os-eol',
-                                    self.feature_set)
+                                    self.feature_change)
 
-    def domain_added(self, _submitter, _event, vm, *_args, **_kwargs):
+    def domain_added(self, _submitter, _event, vmname, *_args, **_kwargs):
```
Comment (marmarek) [inline qui/tray/updates.py:150]: This is wrong, "vm" is used as keyword argument, so you can't change its name. I'll fix it.

I should have put that through openQA earlier...

Comment (alimirjamali) [reply qui/tray/updates.py:150]: Ops. Sorry for mistake. I wonder why it did not show up during the unit tests.  I guess there is no unit test for installing an outdated or obsolete template.

**Pull request merged**